token_with_payload/src/contract.rs

```diff
@@ -0,0 +1,290 @@
+
+    fn approve(
+        e: Env,
+        from: Identifier,
+        nonce: BigInt,
+        spender: Identifier,
+        amount: BigInt,
+        sigs: Map<Identifier, Signature>,
+    );
+
+    fn balance(e: Env, id: Identifier) -> BigInt;
+
+    fn is_frozen(e: Env, id: Identifier) -> bool;
+
+    fn xfer(e: Env, from: Signature, nonce: BigInt, to: Identifier, amount: BigInt);
+
+    fn xfer_from(
+        e: Env,
+        spender: Signature,
+        nonce: BigInt,
+        from: Identifier,
+        to: Identifier,
+        amount: BigInt,
+    );
+
+    fn burn(e: Env, admin: Signature, nonce: BigInt, from: Identifier, amount: BigInt);
+
+    fn freeze(e: Env, admin: Signature, nonce: BigInt, id: Identifier);
+
+    fn mint(e: Env, admin: Signature, nonce: BigInt, to: Identifier, amount: BigInt);
+
+    fn set_admin(e: Env, admin: Signature, nonce: BigInt, new_admin: Identifier);
+
+    fn unfreeze(e: Env, admin: Signature, nonce: BigInt, id: Identifier);
+
+    fn decimals(e: Env) -> u32;
+
+    fn name(e: Env) -> Bytes;
+
+    fn symbol(e: Env) -> Bytes;
+}
+
+fn read_nonce(e: &Env, id: &Identifier) -> BigInt {
+    let key = DataKey::Nonce(id.clone());
+    e.contract_data()
+        .get(key)
+        .unwrap_or_else(|| Ok(BigInt::zero(e)))
+        .unwrap()
+}
+
+fn verify_and_consume_nonce(e: &Env, id: &Identifier, expected_nonce: &BigInt) {
+    match id {
+        Identifier::Contract(_) => {
+            if BigInt::zero(&e) != expected_nonce {
+                panic!("nonce should be zero for Contract")
+            }
+            return;
+        }
+        _ => {}
+    }
+
+    let key = DataKey::Nonce(id.clone());
+    let nonce = read_nonce(e, id);
+
+    if nonce != expected_nonce {
+        panic!("incorrect nonce")
+    }
+    e.contract_data().set(key, &nonce + 1);
+}
+
+pub struct Token;
+
+#[cfg_attr(feature = "export", contractimpl)]
+#[cfg_attr(not(feature = "export"), contractimpl(export = false))]
+impl TokenTrait for Token {
+    fn initialize(e: Env, admin: Identifier, decimal: u32, name: Bytes, symbol: Bytes) {
+        if has_administrator(&e) {
+            panic!("already initialized")
+        }
+        write_administrator(&e, admin);
+
+        write_decimal(&e, u8::try_from(decimal).expect("Decimal must fit in a u8"));
+        write_name(&e, name);
+        write_symbol(&e, symbol);
+    }
+
+    fn nonce(e: Env, id: Identifier) -> BigInt {
+        read_nonce(&e, &id)
+    }
+
+    fn allowance(e: Env, from: Identifier, spender: Identifier) -> BigInt {
+        read_allowance(&e, from, spender)
+    }
+
+    fn approve(
+        e: Env,
+        from_id: Identifier,
+        nonce: BigInt,
+        spender: Identifier,
+        amount: BigInt,
+        sigs: Map<Identifier, Signature>,
+    ) {
+        let from = sigs.get(from_id.clone()).unwrap().unwrap();
+
+        verify_and_consume_nonce(&e, &from_id, &nonce);
+
+        crate::auth::check_auth::<TokenPayload>(
```

Comment:
Is the idea to use/reuse different payloads for different functions that need auth? I suppose you omitted that from the PR, as only this fn uses the new auth.

Reply:
It's exploratory work. I learned what I needed from this part so I didn't do the rest of the work (as noted, I expect everything to change). But in practice every function will use the same new auth.